jicoco/src/main/java/org/jitsi/xmpp/mucclient/MucClient.java

```diff
@@ -857,4 +860,49 @@ private void resetLastPresenceSent()
             lastPresenceSent = null;
         }
     }
+
+    /**
+     * Detects ping failures and in case of two consecutive failures where the connection is still
+     * connected and authenticated we will close it and let the connectRetry task handle that.
+     */
+    private class PingFailedListenerImpl
+        implements PingFailedListener
+    {
+        /**
+         * When ping fails the PingManager stops scheduling new pings.
+         * That's why we will send manually another one to check and disconnect if that fails.
+         */
+        @Override
+        public void pingFailed()
+        {
+            logger.warn("XMPP Ping failed");
+
+            boolean res = false;
+            try
+            {
+
+                PingManager pingManager = PingManager.getInstanceFor(xmppConnection);
+                if (pingManager != null)
+                {
+                    res = pingManager.pingMyServer(false);
+                }
+            }
+            catch (InterruptedException | SmackException e)
+            {
+                res = false;
+            }
+
+            if (!res)
+            {
+                logger.warn("Second XMPP Ping failed");
+
+                if (xmppConnection.isConnected() && xmppConnection.isAuthenticated())
+                {
+                    // two pings failing in a row where connection is still connected and authenticated
+                    // a weired situation, we will trigger reconnect just in case.
+                    xmppConnection.disconnect();
```

Comment:
So, calling `disconnect` trigger another connetion attempt?

Reply:
Yep, cause It hits here https://github.com/jitsi/jicoco/blob/master/jicoco/src/main/java/org/jitsi/xmpp/mucclient/MucClient.java#L289
and uses the connectRetry logic.